File: bubble-sort/bubble-sort.rb
module BubbleSort
  extend self

  def sort(array, order: :asc)
    sorted = true

    case order
    when :asc
      for i in 0..array.length - 2
        if array[i] > array[i + 1]
          sorted = false
          swap(array, i, i + 1)
        end
      end
    when :desc
      for i in 0..array.length - 2
        if array[i] < array[i + 1]
          sorted = false
          swap(array, i, i + 1)
        end
      end
    end

    sorted ? array : sort(array, order: order)
  end

  def swap(array, index_1, index_2)
    temp = array[index_1]
    array[index_1] = array[index_2]
    array[index_2] = temp
  end

end

File: bubble-sort/java/BubbleSort.java
public class BubbleSort {

  public enum Order { ASC, DESC }

  public static int[] sort( int[] array ) {
    boolean sorted = true;
    for (int i = 0; i <= array.length - 2; i++) {
      if (array[i] > array[i + 1]) {
        sorted = false;
        swap(array, i, i + 1);
      }
    }
    return sorted == true ? array : sort(array);
  }

  public static int[] sort( int[] array, Order order ) {
    boolean sorted = true;
    for (int i = 0; i <= array.length - 2; i++) {
      switch (order) {
        case ASC: {
          if (array[i] > array[i + 1]) {
            sorted = false;
            swap(array, i, i + 1);
          }
          break;
        }
        case DESC: {
          if (array[i] < array[i + 1]) {
            sorted = false;
            swap(array, i, i + 1);
          }
          break;
        }
      }
    }
    return sorted == true ? array : sort(array, order);
  }

  public static void swap( int[] array, int index1, int index2 ) {
  	int temp = array[index1];
  	array[index1] = array[index2];
  	array[index2] = temp;
  }

}
